chromaui/chromatic-e2e
# Playwright: Snapshots stored outside of `testInfo.attachments`

Issue: #AP-4453

## What Changed


We now use our custom object (instead of Playwright's `testInfo.attachments`) to store the snapshotted DOM. This is so this DOM won't be captured in reporters, which has caused out-of-memory issues in the [JSON reporter](https://playwright.dev/docs/test-reporters#json-reporter), for example.

### Before
When [`takeSnapshot()`](https://github.com/chromaui/chromatic-e2e/blob/c107353d7abd53037faca212ecc0e9f64c95fa83/packages/playwright/src/takeSnapshot.ts) was called (either [automatically](https://github.com/chromaui/chromatic-e2e/blob/c107353d7abd53037faca212ecc0e9f64c95fa83/packages/playwright/src/makeTest.ts#L79) at the end of the test, or [manually](https://www.chromatic.com/docs/playwright/targeted-snapshots/) by the user in their spec file), we'd [call](https://github.com/chromaui/chromatic-e2e/blob/c107353d7abd53037faca212ecc0e9f64c95fa83/packages/playwright/src/takeSnapshot.ts#L40) Playwright's [`testInfo.attach()`](https://playwright.dev/docs/api/class-testinfo#test-info-attach) to store the DOM snapshot as an attachment. This way we were able to [access the snapshots](https://github.com/chromaui/chromatic-e2e/blob/c107353d7abd53037faca212ecc0e9f64c95fa83/packages/playwright/src/makeTest.ts#L84-L88) later when writing them to disk.

However, this meant that the DOM snapshots (typically very large in size) were now being included in Playwright's reporters. For the JSON reporter, this ballooned the size of the reporter's output file to the point where Playwright was unable to generated the report file (Playwright still finished running the tests).

### Now
Instead of putting the DOM snapshots on Playwright's `testInfo` object as attachments, we [write them](https://github.com/chromaui/chromatic-e2e/blob/721bc9e517d198a505ffcfdd48efb2ff68929b0f/packages/playwright/src/takeSnapshot.ts#L47) to our [own in-memory object](https://github.com/chromaui/chromatic-e2e/blob/721bc9e517d198a505ffcfdd48efb2ff68929b0f/packages/playwright/src/takeSnapshot.ts#L10) (`chromaticSnapshots`), with each test's snapshot(s) being written based on the test's ID. We then [read](https://github.com/chromaui/chromatic-e2e/blob/721bc9e517d198a505ffcfdd48efb2ff68929b0f/packages/playwright/src/makeTest.ts#L84) from `chromaticSnapshots` to get the DOM snapshots to write to disk.

We also [delete the snapshots](https://github.com/chromaui/chromatic-e2e/blob/721bc9e517d198a505ffcfdd48efb2ff68929b0f/packages/playwright/src/makeTest.ts#L105) off the `chromaticSnapshots` object after the test has completed to avoid that object getting massive.

This avoids the DOM snapshots getting written to the Playwright reporters.

## How to test


1. Ensure all Chromatic playwright snapshots (`UI Tests: playwright-e2e` PR status check) are unchanged
3. Clone [this repo](https://github.com/skitterm/playwright-lots-of-tests) and follow the verification steps there (ensures that reporter file is created with this fix, even for large numbers of large tests)
